daemon/mgr/container_copy.go

```diff
@@ -0,0 +1,181 @@
+package mgr
+
+import (
+	"context"
+	"errors"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/docker/docker/pkg/archive"
+	"github.com/docker/docker/pkg/chrootarchive"
+)
+
+// StatPath stats the dir info at the specified path in the container.
+func (mgr *ContainerManager) StatPath(ctx context.Context, name, path string) (stat *ContainerPathStat, err error) {
+	c, err := mgr.container(name)
+	if err != nil {
+		return nil, err
+	}
+	c.Lock()
+	defer c.Unlock()
+
+	err = mgr.Mount(ctx, c)
+	defer mgr.Unmount(ctx, c)
+	if err != nil {
+		return nil, err
+	}
+	err = mgr.attachVolumes(ctx, c)
+	defer mgr.detachVolumes(ctx, c, true)
+	if err != nil {
+		return nil, err
+	}
+
+	// Consider the given path as an absolute path in the container.
+	absPath := path
+	if !filepath.IsAbs(absPath) {
+		absPath = archive.PreserveTrailingDotOrSeparator(filepath.Join("/", path), path, os.PathSeparator)
+	}
+
+	resolvedDirPath := c.GetResourcePath(c.BaseFS, filepath.Dir(absPath))
+
+	resolvedPath := resolvedDirPath + filepath.Base(absPath)
+
+	lstat, err := os.Lstat(resolvedPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ContainerPathStat{
+		Name:  lstat.Name(),
+		Path:  absPath,
+		Size:  lstat.Size(),
+		Mode:  lstat.Mode(),
+		Mtime: lstat.ModTime(),
+	}, nil
+}
+
+// ArchivePath return an archive and dir info at the specified path in the container.
+func (mgr *ContainerManager) ArchivePath(ctx context.Context, name, path string) (content io.ReadCloser, stat *ContainerPathStat, err error) {
+	c, err := mgr.container(name)
+	if err != nil {
+		return nil, nil, err
+	}
+	c.Lock()
+	defer c.Unlock()
+
+	err = mgr.Mount(ctx, c)
+	defer mgr.Unmount(ctx, c)
```

Comment:
Just confirm if there is need to `defer mgr.Unmount(ctx, c)` when the err is not nil.
I see lots of cases like this way below this one.

Reply:
yep, need to unmout it.